### DISCUSSION TITLE
v3 Roadmap

### DISCUSSION THREAD

**Discussion by renanbr** [renanbr/bibtex-parser#96] [General]
#### Goals

- make the code source more modern
- improve code source readability
- improve documentation

#### Topics
- **Entry**: entries as object `Entry` instead of `array`
    - an immutable object
    - do not mix up tags with non-tag stuff, it means provide these methods in `Entry` class:
        - `getCitationKey(): string`
        - `getType(): string`
        - `getOriginal(): string`
        - `getTag(string $name): Stringable` (PHP 8.0)
            - processors may return a "stringable" `ArrayIterator` for lists (e.g. keywords, author)
- **ProcessorInterface**: create a `ProcessorInterface`
    - type-hint processor's argument and return as `Entry` (PHP 7.0)
- **Default Listener**: instantiate a default Listener in Parser
    - provide a way to clear listeners in Parser
- **Abbreviation leak**: don't leak abbreviation among files (bug in v2?)
    - trigger Listeners when Parser reaches the end of a file (or string input)
- **Interfaces namespace**: place interfaces under an unified namespace
- **Static analysis**: stricter static analysis check
- **Documentation**: more examples in documentation
    - how to integrate this library with the most popular frameworks

**Comment by renanbr** [comment]
Topic **Entry** and **ProcessorInterface**:

We could provide a Processor that transforms an array into a `Entry`.

As a consequence, the user that wants to type-hint its processors with `Entry` can do this as well. (no interface in this case)

**Reply by alfsb** [reply]
Instead of an Processor, an `exportEntry()`, or more directly an `parseAsEntries()`, so that new v3 namespaced code can be developed side by side with v2 API. And afterwards, another functions that directly calls into v3 code and export the parsed file as XML/JSON/[S-expression](https://en.wikipedia.org/wiki/S-expression), and possibly v2 code can also ex

**Comment by renanbr** [comment]
Topic **Default Listener**

We could provide a static method anywhere as a factory, or even a `ParserBuilder`.

**Comment by renanbr** [comment]
Topic **Refactor**

Extract lexer from this project.

**Reply by alfsb** [reply]
Making separated lexer classes, that conforms to an lexer interface, allows to "configure" the new v3 code into something way more easy to configure. The https://github.com/renanbr/bibtex-parser/pull/112 provides a good example. Having the parser calling into replaceable dependency injected lexers, allow to have effective different parsers without changing the parser class to much, or having a lot of knobs.
